### PR TITLE
fix(seeder): detect truncated utxo-headers files in the header read loops

### PR DESCRIPTION
**Pull request by icellan**
Split out of #1572 per review feedback — this is a real behavioral fix, not test-only, and deserves review on its own terms.

The same `errors.Is(err, io.EOF)` substring-matching hazard that #1563 fixes for UTXO-set files also exists, unpatched, in the utxo-headers read loops in `cmd/seeder/seeder.go`: a genuinely truncated headers file (cut mid-record) produces `io.ErrUnexpectedEOF`, which this codebase's custom `(*Error).Is` matches against a clean `io.EOF` via substring fallback ("EOF" is a substring of "unexpected EOF") — so truncation was silently treated as a clean end-of-records.

Fixed the headers loop to distinguish genuine truncation from a clean footer boundary. Note: #1563, which applies the equivalent fix to the UTXO-set read loop, is still open — the two loops are independent and neither depends on the other landing first. Preserved the existing best-effort partial-load behavior for the coinbase-loading caller (returns the partial map alongside the error, rather than discarding everything already loaded on any truncation) so a headers file truncated at 90% still yields ~90% of coinbase transactions, with the warning now firing correctly on real truncation instead of never firing at all.

Follow-up from review:
- `processUTXOs` no longer silently reports success when `lastProcessed.dat` exists but no BlockAssembler checkpoint does — that combination means an earlier run imported the UTXO set but failed before the checkpoint was written (e.g. this fix's own header-truncation check rejecting the file after the UTXO pass had already finished). It now refuses and points at `-force`.
- `loadCoinbaseTxs`'s non-EOF read/decode-error path also now returns the partial coinbase map instead of discarding it, matching the record-count-mismatch path.
- Doc comments and error/warning text corrected for accuracy.

### CODE REVIEW

**Review by ctnguyen — CHANGES_REQUESTED** [review]
Review at commit [33ededa](https://github.com/icellan/teranode/commit/33ededa9bb52b28eaf5f7d6cbb90759a570b39c9)

Splitting this out of #1572 was the right call — it's a real behavioural fix and it reads much better on its own. The diagnosis is correct and I re-derived it from the code: `NewUTXOHeaderFromReader` wraps every short read in `errors.NewProcessingError` (`services/utxopersister/Headers.go:142`), `(*Error).Error()` appends a non-`*Error` tail verbatim (`errors/errors.go:137`) so both `io.EOF` and `io.ErrUnexpectedEOF` render with "EOF" in the message, and `(*Error).Is` substring-matches non-`*Error` targets (`errors/errors.go:177`) — the two really were indistinguishable at `seeder.go:329`. The choice of ground truth is also right: the preamble tip height sits *before* any truncation point, so a truncated file can't forge it, and `WriteHeadersToStore` (`services/utxopersister/UTXOSet.go:875-946`) does walk tip→genesis and stop at `meta.Height == 0`, emitting one record per height — the shipped `seeds/tstn` snapshot (tip height 300, "headers up to and including the tip block") matches that shape. Validating a count instead of touching the file format keeps the diff small, and the positive-path tests guarding against the new check rejecting a legitimate snapshot are a nice touch. Two items below are worth fixing before this lands; the rest are smaller.

## Blocking issues

### ChiR1 — The new hard failure leaves a half-seeded store whose re-run silently skips the BlockAssembler checkpoint

**Problem:** `processHeaders` and `processUTXOs` run as independent goroutines (`cmd/seeder/seeder.go:191-234`) and `headerErr` is only inspected after `wg.Wait()`. A header failure does not cancel the UTXO pass, which runs to completion and writes the `lastProcessed.dat` marker at `cmd/seeder/seeder.go:596`. Because a truncated headers file is exactly the case where the `.utxo-set` file is usually intact (separate file, separate copy), the realistic sequence after this PR is:

1. Run 1: headers fail with the new "utxo-headers file truncated" error; UTXOs import fully; `lastProcessed.dat` is written; `Seeder` returns an error at line 238-240, so `writeBlockAssemblerState` (line 254-258) is skipped.
2. Operator re-copies the headers file and re-runs. `processUTXOs` sees `lastProcessed.dat` and returns `(nil, nil)` at line 424-430 — before it ever reaches the `blockAssemblerStateExists` check at line 437.
3. `utxoTip` is `nil`, so the guard at line 254 skips the checkpoint write, and `Seeder` returns `nil`.

The seeder now reports success on a store that has headers, has the full UTXO set, and has no BlockAssembler checkpoint — step 3 of the documented import (`seeds/tstn/README.md:31-33`) silently never happened.

**Why it matters:** This is the failure the code already calls out as unacceptable in its own comment at `cmd/seeder/seeder.go:246-253` ("the node would otherwise start with no checkpoint and a plain re-run skips the (now-complete) UTXO pass, never retrying it"). Before this PR a truncated headers file returned `nil`, so this path was not reachable from truncation; the fix converts one silent mis-seed into a different one that is arguably worse, because run 2 exits 0. Recovery requires a full `-force` UTXO re-import, which on a real snapshot is expensive and is the flag the README warns against. Given that the point of this PR series is to make a truncated snapshot loud instead of silent, a path that ends in a false success shouldn't ship alongside it.

**Fix:** `lastProcessed.dat` existing without a checkpoint is precisely the inconsistent half-seeded state, so detect it and either recover or refuse — don't return `nil`. The utxo-set preamble already carries the tip hash and height, so recovery is cheap:

```go
if exists {
    // lastProcessed.dat without a BlockAssembler checkpoint means a previous
    // run imported the UTXO set but failed before the checkpoint was written
    // (e.g. the header pass rejected a truncated file). Recover the tip from
    // the utxo-set preamble so the re-run completes the seed instead of
    // silently reporting success with no checkpoint.
    var stateExists bool

    if stateExists, err = blockAssemblerStateExists(ctx, blockchainStore); err != nil {
        return nil, errors.NewStorageError("failed to check BlockAssembler state", err)
    }

    if !stateExists {
        // new helper: reads the tip hash + height from the utxo-set preamble
        return readUTXOSetTip(utxoFile)
    }

    logger.Infof("lastProcessed.dat exists, skipping UTXOs")

    return nil, nil
}
```

If you'd rather not add tip recovery here, the minimum is to return an error from that branch when no checkpoint exists, naming `-force` as the recovery — a loud failure is acceptable, a silent `exit 0` is not.

### ChiR2 — `loadCoinbaseTxs` still discards the partial map on the decode/IO-error path

**Problem:** The count-mismatch branch at `cmd/seeder/seeder.go:817` returns the partial map. The branch three lines earlier does not:

```go
// cmd/seeder/seeder.go:788-795
blockIndex, err = utxopersister.NewUTXOHeaderFromReader(reader, false)
if err != nil {
    if errors.Is(err, io.EOF) {
        break
    }

    return nil, errors.NewProcessingError("failed to read UTXO header", err)  // <- discards coinbaseTxs
}
```

I traced which failures actually land there, because it is narrower than it looks. Pure truncation never does: every read in `NewUTXOHeaderFromReader` goes through `io.ReadFull`, so a missing tail always surfaces as `io.EOF` or `io.ErrUnexpectedEOF`, both of which render with "EOF" in the wrapped message and therefore satisfy `errors.Is(err, io.EOF)` and `break`. The 80-byte block-header parse can't fail either — `NewBlockHeaderFromBytes` only errors on a wrong length or a bad 32-byte slice (`model/BlockHeader.go:74-99`), neither of which is possible after a successful `ReadFull` of exactly 80 bytes. What does reach line 794 is:

- an underlying read error that isn't EOF — a failing disk or a network filesystem dropping out mid-file, which is one of the transport faults #1563's rationale calls out for large snapshots;
- a corrupted coinbase-length field or corrupted coinbase bytes, where `coinbaseLen` bytes are read in full but `bt.NewTxFromBytes` rejects them (`services/utxopersister/Headers.go:194-197`).

In both cases every coinbase read before the bad record is thrown away, and the caller's `if coinbaseTxs == nil` guard at `cmd/seeder/seeder.go:469` substitutes an empty map, so the loss is invisible.

**Why it matters:** To be precise about scope: the PR description's claim is about truncation, and for truncation it holds — I'm not saying the stated fix is false. The defect is that `loadCoinbaseTxs` now has two error exits with opposite contracts and nothing tells them apart, so identical-looking damage to a headers file yields "90% of coinbases restored" or "none restored" depending on which byte got corrupted. That's not something a caller can reason about, and the discarding exit is the one nothing in the code or doc comment mentions. It's a one-line fix, so there's no reason for the two paths to disagree.

**Fix:** Return the partial map from the decode/IO-error path too, for the same reason the count-mismatch path does.

```go
if err != nil {
    if errors.Is(err, io.EOF) {
        break
    }

    // Same rationale as the record-count mismatch below: the caller treats any
    // error as "restore no coinbase inputs at all", so hand back what was read.
    return coinbaseTxs, errors.NewProcessingError("failed to read UTXO header", err)
}
```

The `if coinbaseTxs == nil` guard at line 469 should stay — the open and header-validation paths legitimately return `nil` — but it currently hides this case rather than documenting it.

## Non-blocking issues

### ChiR3 — Nothing in the new tests pins the discard path from ChiR2

**Problem:** `TestLoadCoinbaseTxs_TruncatedFile_ReturnsPartialMap` (`cmd/seeder/seeder_headers_truncation_test.go:122-181`) is the only test asserting the partial-map contract, and it builds its damage as `block2Record[:10]` (line 169) — a mid-record cut. As traced in ChiR2, that fails inside an `io.ReadFull`, breaks out of the loop, and is caught by the record-count check, so the test exercises only the count-mismatch exit. `TestLoadCoinbaseTxs_TruncatedFile_ReturnsError` (line 102-110) uses the same construction. The decode/IO-error exit at `seeder.go:794` — the one that still returns `nil` — has no test at all, in either direction.

**Why it matters:** Once ChiR2 is fixed, the fix will be unprotected: a later cleanup restoring the idiomatic `return nil, err` there would pass the whole suite. It also means the current suite cannot demonstrate the gap, which is part of why it's easy to miss.

**Fix:** Add a case whose final record is read in full but fails to decode, rather than being cut short — for example a record with a valid 124-byte fixed prefix followed by a coinbase length of 4 and four bytes that are not a serialisable transaction, so `bt.NewTxFromBytes` rejects them. Assert on the map contents rather than the error text, so the test survives message changes.

```go
// a fully-present but undecodable final record: reaches the non-EOF exit,
// unlike the mid-record cut used by the truncation tests above. Copy rather
// than append onto the slice, so block2Record's backing array is untouched.
// 124 = 32 (hash) + 4 (height) + 8 (txCount) + 80 (block header)
bad := append([]byte{}, block2Record[:124]...)
bad = append(bad, 0x04, 0x00, 0x00, 0x00, 0xde, 0xad, 0xbe, 0xef)

path := writeUtxoHeadersFile(t, 2, genesisRecord, block1Record, bad)

coinbaseTxs, err := loadCoinbaseTxs(ulogger.TestLogger{}, path)
require.Error(t, err)
require.Contains(t, coinbaseTxs, *coinbase1.TxIDChainHash(),
    "coinbases read before an undecodable record must survive, as they do on truncation")
```

### ChiR4 — Both new comments point at `readUTXOWrapperFile`, which does not exist in this tree

**Problem:** `cmd/seeder/seeder.go:372` and `cmd/seeder/seeder.go:806` both say "see the doc comment on `readUTXOWrapperFile` for why". There is no such identifier anywhere in the repository at this commit — it is introduced by #1563, which is still open. Relatedly, the PR description says the fix mirrors "the fix #1563 already applied to the UTXO-set loop", but on `main` the utxo-set read loop at `cmd/seeder/seeder.go:557-565` is still the raw `errors.Is(err, io.EOF)` form with no footer check, even though `utxopersister.GetFooter` exists at `services/utxopersister/Footers.go:42`.

**Why it matters:** The stated reason for splitting this out of #1572 was that it should stand on its own terms — but as written it reads as if it depends on a sibling PR that hasn't landed. If this merges first, a reader chasing the rationale finds nothing. It also makes the merge order load-bearing in a way nothing in the PR states (see [bsv-blockchain/teranode#1563](https://github.com/bsv-blockchain/teranode/pull/1563)).

**Fix:** Make the comments self-contained rather than cross-referencing an unmerged symbol, and note the ordering in the PR description if you intend #1563 to land first.

```go
// The read loop above only knows the record stream ended - it cannot tell a
// clean end-of-records boundary apart from a file truncated mid-record: a short
// read surfaces as io.ErrUnexpectedEOF, which (*errors.Error).Is matches against
// io.EOF via its substring fallback ("EOF" is a substring of "unexpected EOF").
```

### ChiR5 — The `(partial map, error)` contract lives in a body comment, not on the function

**Problem:** `loadCoinbaseTxs`' doc comment (`cmd/seeder/seeder.go:725-728`) still describes only the pre-PR behaviour. The new, deliberately unusual contract — a non-nil, usable map returned *alongside* a non-nil error — is explained only in a comment inside the `if` block at lines 811-816, where a caller doesn't look.

**Why it matters:** Go's default reading of `(T, error)` is that `T` is meaningless when `err != nil`. The next caller, or a lint-driven cleanup, writing the idiomatic `if err != nil { return nil, err }` silently reintroduces the regression this PR just removed, and no test at the `processUTXOs` level would catch it — the new test only pins `loadCoinbaseTxs` itself.

**Fix:** State the contract on the function so it survives refactoring.

```go
// loadCoinbaseTxs reads the V2 utxo-headers file and returns a map from coinbase
// txid to the coinbase transaction. [...]
//
// Contract: on error the returned map is still usable and holds every coinbase
// read before the failure - callers must keep it rather than discarding it,
// since coinbase input restoration is best-effort and a damaged file should
// degrade to "most coinbases restored", not "none". The map is nil only when the
// file could not be opened or its header failed validation.
```

### ChiR6 — The truncation tests assert only "some error", and miss the case the count check uniquely catches

**Problem:** `require.Error(t, err, ...)` at `cmd/seeder/seeder_headers_truncation_test.go:97`, `:109` and `:172` doesn't pin *which* error fired, so the tests would keep passing if the new validation were removed and something else in the path started failing instead. More substantively, all three cases cut mid-record (`secondRecord[:10]`), which is the `io.ErrUnexpectedEOF` flavour. The case the record-count check uniquely catches — a file cut *exactly* on a record boundary, which produces a genuinely clean `io.EOF` and is indistinguishable by any error-inspection fix — is untested. There's also no coverage for a complete V1 file, even though `processHeaders` still supports V1 (`cmd/seeder/seeder.go:317`) and the new check now hard-fails V1 files on the same invariant; `fileformat.NewHeader(FileTypeUtxoHeaders)` resolves to the V2 magic (`pkg/fileformat/header.go:78`), so every new test is V2.

**Why it matters:** The boundary-aligned cut is the strongest argument for validating a count rather than untangling the sentinel plumbing — it deserves to be the test that pins the design in place. And a V1 regression would only show up on an operator's legacy snapshot.

**Fix:** Tighten the assertions and add the boundary case; a V1 positive-path case is cheap once `writeUtxoHeadersFile` can emit a V1 magic.

```go
require.ErrorContains(t, err, "truncated",
    "the record-count validation must be what rejects the file")

// boundary-aligned cut: no partial record at all, so the read loop sees a clean io.EOF
path = writeUtxoHeadersFile(t, 1, genesisRecord)
require.ErrorContains(t, processHeaders(ctx, ulogger.TestLogger{}, store, path), "truncated")
```

One easy win alongside this: `TestLoadCoinbaseTxs_RealSeedFile` (`cmd/seeder/coinbase_restore_test.go:112-117`) points at `000000007816d70a…utxo-headers`, but the snapshot committed under `seeds/tstn/` is `000000000c130c0b…`, so it always takes the `t.Skipf` branch. Repointing it exercises the new invariant against the real shipped snapshot for free.

### ChiR7 — Two of the new messages describe the wrong thing

**Problem:** Both new errors are worded as "utxo-headers file truncated" (`cmd/seeder/seeder.go:380`, `:818`) but fire on any inequality, including `recordsRead > tipHeight+1` — a file with extra or duplicated records, which is not truncation. Separately, the warning at `cmd/seeder/seeder.go:467` was softened from "coinbases will be stored without their input" to "some coinbases may be stored without their input", but it fires for every `loadCoinbaseTxs` error, including the paths that return `nil` (open failure at line 743, header read failure at 754, wrong file type at 758, preamble read failure at 773/778, and the decode path in ChiR2) — in all of those, *no* coinbases are restored.

**Why it matters:** These messages are what an operator sees at seed time on a snapshot they need to decide whether to trust. "Some may be missing" reads as a minor degradation when the actual outcome can be that every coinbase is stored in the output-only form.

**Fix:** Say "record count mismatch" and let the numbers carry the direction, and keep the warning honest about the total-loss case.

```go
return errors.NewProcessingError(
    "utxo-headers file record count mismatch: expected %d records (heights 0..%d), read %d",
    uint64(height)+1, height, recordsRead)

logger.Warnf("[processUTXOs] could not fully load coinbase transactions from %s; %s coinbases recovered, the rest will be stored without their input: %v",
    headersFile, formatNumber(uint64(len(coinbaseTxs))), err)
```

## Recap

| ID | Description | Required | Criticality |
| --- | --- | --- | --- |
| ChiR1 | Re-run skips assembler checkpoint | YES | 72% |
| ChiR2 | Partial map dropped on decode error | YES | 62% |
| ChiR3 | Discard path untested | | 45% |
| ChiR4 | Comments cite non-existent function | | 40% |
| ChiR5 | Partial-map contract undocumented | | 35% |
| ChiR6 | Weak assertions, missing cases | | 25% |
| ChiR7 | Misleading error and warning text | | 18% |

**Comment by icellan** [comment]
Pushed 413a96ee0 addressing the review.

- **ChiR1 (blocker)**: extracted the skip/refuse decision into `checkSkipUTXOImport`. It now checks the BlockAssembler checkpoint in both branches: `lastProcessed.dat` present with no checkpoint refuses with an error naming `-force`, instead of silently returning `(nil, nil)` and letting the caller skip the checkpoint write. Went with the "minimum" option from the review (loud failure) rather than recovering the tip from the utxo-set preamble, to keep the diff to the actual bug. Covered by `TestCheckSkipUTXOImport_MarkerWithoutCheckpoint_RefusesWithForceHint` plus the no-marker and clean-skip cases in `seeder_state_test.go`.
- **ChiR2 (blocker)**: `loadCoinbaseTxs`'s non-EOF read/decode-error exit now returns `coinbaseTxs, err` instead of `nil, err`, matching the record-count-mismatch path.
- **ChiR3**: added `TestLoadCoinbaseTxs_UndecodableRecord_ReturnsPartialMap` — a fully-present final record with an undecodable coinbase, reaching the ChiR2 exit rather than the mid-record-cut/EOF path the existing tests hit.
- **ChiR4**: both dangling `readUTXOWrapperFile` comments replaced with the actual rationale (`errors.Error.Is` substring fallback). PR description corrected — it no longer claims #1563 already covers the UTXO-set loop (still open, independent fix).
- **ChiR5**: the `(partial map, error)` contract is now on `loadCoinbaseTxs`'s doc comment, not just inline in the `if` body.
- **ChiR6**: tightened `require.Error` to `require.ErrorContains(..., "truncated")` in the existing truncation tests; added a boundary-aligned-cut case (`TestProcessHeaders_BoundaryAlignedCut_ReturnsError`, clean `io.EOF`, only the record-count check catches it) and a V1-file positive-path case (`TestProcessHeaders_V1File_FinishesCleanly`). Also fixed `TestLoadCoinbaseTxs_RealSeedFile`, which pointed at a hash not present under `seeds/tstn/` and always skipped — repointed at the actual committed seed file and verified the block-300 coinbase txid against it.
- **ChiR7**: reworded both messages to lead with "record count mismatch (likely truncated)" rather than asserting truncation for a too-many-records case, and made the coinbase-restoration warning report the actual recovered count rather than implying partial recovery on paths where none occurred.

`go build ./...`, `go vet ./cmd/seeder/...`, `go test ./cmd/seeder/... -race` all pass.

File: cmd/seeder/seeder.go [hunk 1]
```diff
@@ -364,6 +367,20 @@ func processHeaders(ctx context.Context, logger ulogger.Logger, blockchainStore
 		}
 	}
 
+	// The read loop above only knows the record stream ended - it cannot tell
+	// a clean end-of-records boundary apart from a file truncated mid-record
+	// (see the doc comment on readUTXOWrapperFile for why). Unlike utxo-set
```
[the remark this answers is not in the document]
Comment (icellan) [reply cmd/seeder/seeder.go]: Fixed in 413a96ee0 — both comments now inline the errors.Error.Is substring-matching rationale instead of pointing at readUTXOWrapperFile.

File: cmd/seeder/seeder.go [hunk 2]
```diff
@@ -364,6 +367,20 @@ func processHeaders(ctx context.Context, logger ulogger.Logger, blockchainStore
...
+	// files, utxo-headers files carry no trailing footer, but WriteHeadersToStore
```
[the remark this answers is not in the document]
Comment (icellan) [reply cmd/seeder/seeder.go]: Fixed in 413a96ee0 — both comments now inline the errors.Error.Is substring-matching rationale instead of pointing at readUTXOWrapperFile.

File: cmd/seeder/seeder.go [hunk 3]
```diff
@@ -364,6 +367,20 @@ func processHeaders(ctx context.Context, logger ulogger.Logger, blockchainStore
...
+	// always writes exactly one record per height from genesis (0) up to the
+	// tip height already read above, so that tip height is an independent
+	// ground truth we can validate the read count against without any file
+	// format change.
+	if recordsRead != uint64(height)+1 {
+		return errors.NewProcessingError(
+			"utxo-headers file truncated: expected %d header records (heights 0..%d), read %d",
+			uint64(height)+1, height, recordsRead)
+	}
```
[the remark this answers is not in the document]
Comment (icellan) [reply cmd/seeder/seeder.go:383]: Fixed in 413a96ee0 — both messages now include headersFile.

**Review by ordishs — CHANGES_REQUESTED** [review]
**cmd/seeder/seeder.go:869** — P1: the prescribed `-force` recovery cannot complete the seed.

The new refusal tells the operator to "re-run with `-force` to re-import the UTXO set and complete the seed". No invocation reaches a written checkpoint from that state. Two independent blockers:

- `-force` only reaches `processUTXOs` (`seeder.go:222`). A re-run still runs `processHeaders`, which calls `StoreBlock` for blocks already stored. `stores/blockchain/sql/StoreBlock.go:652` returns `BlockExistsError`, `seeder.go:358-360` makes that fatal, and `Seeder` returns at `seeder.go:238-240` — before `writeBlockAssemblerState` at `seeder.go:254`.
- Adding `-skipHeaders` clears that, but the marker rewrite at `seeder.go:584` calls `blockStore.Set` with no `WithAllowOverwrite`. `AllowOverwrite` defaults to false (`stores/blob/options/Options.go:42-43`) and `stores/blob/file/file.go:646-654` returns `BlobAlreadyExists`. The default blockstore is `file://` in every context (`settings.conf:390-393`). So the forced pass re-imports the whole UTXO set — hours on mainnet — then fails on the final write and leaves the store in the same refusing state.

Reproduction: state is `lastProcessed.dat` present with no checkpoint, which is exactly what this PR's own truncation check produces when the UTXO pass finishes before the header pass rejects the file. `seeder … -force` fails with `BLOCK_EXISTS (12)`. `seeder … -force -skipHeaders` fails with `BLOB_EXISTS` after a full re-import. The only working recovery — delete `lastProcessed.dat`, then re-run — is not mentioned.

Suggested fix:
- Pass `bloboptions.WithAllowOverwrite(true)` on the marker write at `seeder.go:584`, with a test that forces over a store already holding `lastProcessed.dat`.
- Correct the message to name the real recovery, or make `processHeaders` treat `errors.ErrBlockExists` as a skip so `-force` alone works.
- Cheaper still: the UTXO-set preamble already yields the tip hash and height (`seeder.go:507-516`, returned at `seeder.go:588`), so the missing checkpoint could be written without re-importing anything.

Everything else checks out. The premise is confirmed — `errors/errors.go:173-177` does substring-match a non-`*Error` target, so "unexpected EOF" matched `io.EOF` — and the count-against-tip-height approach is stronger than an `io.ErrUnexpectedEOF` test, since it also catches truncation on an exact record boundary. The invariant holds for both producers (`services/utxopersister/UTXOSet.go:867-955` and `cmd/bitcointoutxoset/bitcoin/headers.go:104-262`), and the committed testnet file satisfies it (301 records, tip height 300).

**Comment by icellan** [comment]
Addressed the P1: the -force recovery now actually reaches a written checkpoint.

Implemented the third option (cheapest) combined with the first, per your suggestion:

- `checkSkipUTXOImport`'s no-checkpoint state (`seeder.go:868`, `-force` branch) no longer re-imports the UTXO set. Since `lastProcessed.dat` presence is itself proof the prior UTXO import completed, the tip hash/height is instead read straight from the existing `.utxo-set` file's preamble (`readUTXOSetTip`/`readUTXOSetPreamble`) and handed to `writeBlockAssemblerState` - no re-import, no wait.
- `processHeaders` now treats `errors.ErrBlockExists` as an idempotent skip rather than fatal, so a `-force` re-run doesn't die on blocks a prior pass already stored (this still matters for the case where headers genuinely need completing before the checkpoint's tip header is resolvable).
- The marker rewrite at the end of `processUTXOs` now passes `bloboptions.WithAllowOverwrite(true)`, so that write can't itself be the reason a recovery fails.
- The refusal message no longer promises "re-import the UTXO set"; it now says `-force` recovers the checkpoint from the existing import, no re-import needed.

Reproduced both original failures before fixing (`BLOCK_EXISTS (12)` on header re-run, `BLOB_EXISTS` on the marker rewrite) and added regression tests: `cmd/seeder/seeder_checkpoint_recovery_test.go`.

`go build ./...`, `go vet ./cmd/seeder/...`, `go test ./cmd/seeder/... -race`, and `golangci-lint run cmd/seeder/...` all pass.

**Comment by icellan** [comment]
Rebased onto current origin/main (previously conflicting).

Only one conflict: `cmd/seeder/seeder.go`, in the last commit (`fix(seeder): make the -force checkpoint recovery actually complete`). It was an add/add conflict — main's recent history added `readUTXOWrapperFile` at the same spot this branch added `readUTXOSetPreamble`/`readUTXOSetTip`/`checkpointRecoverable`. Both are unrelated, non-overlapping functions, so resolved as a union (kept both). Git's auto-merge had dropped the closing `}` for `readUTXOWrapperFile` with no conflict marker around it; caught and restored before building.

Verified: `go build ./...`, `go vet ./cmd/seeder/...`, and `go test ./cmd/seeder/... -race` all pass. The checkpoint-recovery path, the `WithAllowOverwrite(true)` marker write, and `processHeaders`' `ErrBlockExists` idempotent-skip are all intact post-rebase.

File: cmd/seeder/seeder.go [hunk 4]
```diff
@@ -819,6 +957,46 @@ func newBlockchainStore(logger ulogger.Logger, appSettings *settings.Settings) (
...
+	if !exists {
+		if stateExists {
+			return false, errors.NewProcessingError("BlockAssembler state already exists in the blockchain store — refusing to seed UTXOs over a store already owned by a block assembler; use -force to override")
+		}
+
+		return false, nil
+	}
+
+	if !stateExists {
+		return false, errors.NewProcessingError("lastProcessed.dat exists but no BlockAssembler checkpoint was found in the blockchain store — a previous run likely imported the UTXO set but failed before the checkpoint was written; re-run with -force to recover the checkpoint from the existing UTXO import (no UTXO re-import is needed)")
```
[the remark this answers is not in the document]
Comment (icellan) [reply cmd/seeder/seeder.go]: Good catch on the gap between what the message promises and what -force actually achieves in that state. Fixed the hint in `410bf8747`: it now says that if the prior run failed on a truncated headers file, -force alone hits the same header check again and the headers file has to be replaced with a complete one first.

Deliberately did not offer -skipHeaders as the shortcut. It would get past the record-count check, but your own reasoning is why it should not be suggested — the header pass has to re-store block H for the recovered checkpoint to be valid, so skipping it trades a clean failure for a quietly wrong one.

**Review by ordishs — CHANGES_REQUESTED** [review]
**cmd/seeder/seeder.go:437-453** — `-force` checkpoint recovery trusts a snapshot-agnostic marker

- `checkpointRecoverable` only tests `lastProcessed.dat` for existence. The marker is written with `WithNoHashPrefix` at a fixed path, so it proves *an* import finished — not that it was this `utxoFile`.
- Reachable sequence: seed from snapshot A (tip height 800000) → the header pass rejects A's truncated `utxo-headers` file → marker written, no checkpoint. The new error at `cmd/seeder/seeder.go:994` directs the operator to replace the headers file with a complete one. They fetch a fresh snapshot B (tip height 810000) and re-run with `-force`. `checkpointRecoverable` returns true, no UTXO import runs, and the BlockAssembler checkpoint is committed at B's tip while the UTXO store still holds A's set. Block assembly then resumes 10000 blocks ahead of the UTXO set it builds on, with no warning.
- The pre-PR code did a full re-import in this state, so this is a regression introduced here.
- Fix: read the marker's contents rather than testing for existence — `services/utxopersister/Server.go:766` already has the pattern. Parse the height and compare it with the `utxo-set` preamble height from `readUTXOSetPreamble`. Take the cheap recovery only on an exact match; on a mismatch fall through to the full re-import, which also restores `-force`'s original "actually re-import" meaning.

The truncation fix itself is sound. Validating the record count against the preamble tip height is the right ground truth given `utxo-headers` files carry no footer, `WriteHeadersToStore` guarantees one record per height from 0 to tip, and the check also catches the boundary-aligned cut that no error-inspection fix could. Test coverage of that part is strong.

Lower-priority notes (P2/P3) held back: the `WithAllowOverwrite` comment at seeder.go:543 names a path the new control flow makes unreachable; `TestProcessHeaders_ReRun_FailsOnAlreadyStoredBlock` asserts `NoError`, the opposite of its name; the count check at seeder.go:375 fires after blocks are already persisted, which the error text does not mention. Happy to post those too if useful.

**Comment by icellan** [comment]
Fixed in `7beb3dfe7`, and you were right that this was a regression this PR introduced rather than a pre-existing gap.

`checkpointRecoverable` now reads the marker instead of testing for it. The marker was never existence-only — `processUTXOs` writes `fmt.Sprintf("%d\n", height)` — so the height was there to check all along, and `services/utxopersister/Server.go` `readLastHeight` already parses it exactly that way. Reused that shape: `Get` with the same blob options, `TrimSpace` + `ParseUint`, then compare against `readUTXOSetTip(utxoFile)`. Only an exact match takes the cheap path; a mismatch falls through to the full re-import, which restores what -force used to mean.

Test encodes your sequence directly — marker at 800000, `utxo-set` preamble at 810000, asserts not recoverable. Being straight about the TDD here: the first run failed to compile rather than failing an assertion, because the function did not take `utxoFile` at all, so the case could not even be expressed against the old signature. A real red, but a compile error, not a runtime one.

Checked the sibling path while in there: `checkSkipUTXOImport` is not susceptible to the same conflation, because it only skips when the marker and the checkpoint both exist, and the checkpoint is written by the same run that wrote the marker. Left it alone.

All three of the P2/P3 notes taken:

- The `WithAllowOverwrite` comment described a scenario that cannot reach that line (a full match returns early). Rewritten to describe what actually reaches it: marker absent, mismatched, or matched-with-existing-checkpoint.
- `TestProcessHeaders_ReRun_FailsOnAlreadyStoredBlock` — the assertion and the doc comment were right and the name was wrong, so it is now `..._IsIdempotentOnAlreadyStoredBlock`, with the comment saying it *would have* failed pre-fix rather than implying it still does.
- The count-check error now says the blocks read so far are already persisted, using the existing count. Deliberately did not touch the identical-looking message in `loadCoinbaseTxs` — that loop persists nothing, so the same wording would be false there.

`go build ./...`, `go vet ./cmd/seeder/...`, `go test -race ./cmd/seeder/...` and `golangci-lint run cmd/seeder/...` all clean.
